### ISSUE TITLE
Path::is_trivial

### ISSUE THREAD

**Issue by lolbinarycat** [rust-lang/libs-team#487]
# Proposal

## Problem statement

The standard library currently contains functions that distinguish between relative and absolute paths, but sometimes it is desirable to only accept an even more restricted subset of paths.

## Motivating examples or use cases

Programs that do filesystem traversal within an inner loop may want to reuse a buffer to store paths in, but this can lead to logic errors if those paths are complex, such as `C:something` or `../foo`.

This could also be useful for programs that want to validate a configuration file or path-like argument, for example git would want to reject branch names containing `..`, or that would be stored on another drive.

## Solution sketch

```rust
impl Path {
  /// Tests if joining this path with another path will simply append the two.
  ///
  /// Some things that would prevent a path from being counted as trivial:
  /// * being an absolute path
  /// * windows drive letters
  /// * any `..` section
  /// * windows verbatim paths
  pub fn is_trivial(&self) -> bool;
}
```

## Alternatives
* add a plethora of fine-grained methods for various properties, such as `has_drive_letter` and `is_verbatim`.

## What happens now?

This issue contains an API change proposal (or ACP) and is part of the libs-api team [feature lifecycle]. Once this issue is filed, the libs-api team will review open proposals as capability becomes available. Current response times do not have a clear estimate, but may be up to several months.

[feature lifecycle]: https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html

## Possible responses

The libs team may respond in various different ways. First, the team will consider the *problem* (this doesn't require any concrete solution or alternatives to have been proposed):

- We think this problem seems worth solving, and the standard library might be the right place to solve it.
- We think that this probably doesn't belong in the standard library.

Second, if there's a concrete solution:

- We think this specific solution looks roughly right, approved, you or someone else should implement this. (Further review will still happen on the subsequent implementation PR.)
- We're not sure this is the right solution, and the alternatives or other materials don't give us enough information to be sure about that. Here are some questions we have that aren't answered, or rough ideas about alternatives we'd want to see discussed.

**Comment by kennytm** [comment]
Is *this* definition of "trivial" used anywhere else? I think it may be better to break it down into two functions that has clearer definitions

1. The path is not absolute, has no drive letter, etc (exists today already as `!path.has_root()`)
2. The path contains no `..` interior (based on #396 I'd call it `is_normalized_lexically()`):

```rust
fn is_trivial(path: &Path) -> bool {
    !path.has_root() && path.is_normalized_lexically()
}
```

Questions: are these considered trivial / normalized lexically?

1. `dir/`
2. `./file`
3. `dir//////file`
4. `dir/././././file`
5. `.`
6. the empty path `""`

> for example git would want to reject branch names containing `..`, or that would be stored on another drive.

for this motivation the proposed function based on Path may be inappropriate when portability is considered, for instance on Unix you a path `x\..\y` is definitely "trivial", but a git repo with such branch name may spell trouble for programs relying on `is_trivial()` when run on Windows.

**Comment by ChrisDenton** [comment]
> The path is not absolute, has no drive letter, etc (exists today already as `!path.has_root()`)

That will return true for both `file.txt` and for `C:file.txt` paths. I think a proper function would be more like:

```rust
fn has_windows_legacy_prefix(path: &Path) -> bool {
    !path.is_absolute() && matches!(path.components().next(), Some(Component::Prefix(_)))
}
```

Then `!has_windows_legacy_prefix` would work.

**Comment by kennytm** [comment]
@ChrisDenton Ok :(

For the first test I think what we are interested in is whether the path is "completely relative". In [Python](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.anchor) terms meaning the path has no "anchor"

```rust
impl Path {
    pub fn is_anchored(&self) -> bool {
        matches!(self.components().next(), Some(Component::Prefix(_) | Component::RootDir)))
    }
}
```

**Comment by ChrisDenton** [comment]
I would propose a `RelativePath` type for that rather then relying on checking. We could have a path that works the same cross-platform (give or take path separators perhaps) and not have any of the funky platform specific functions (or if we do then only as platform-specific extensions).

**Comment by lolbinarycat** [comment]
> I would propose a `RelativePath` type for that rather then relying on checking.

you still need to have the checking if you ever want to take it from user input.

also, `c:foo` is a trivial path on linux but not on windows.

**Comment by ChrisDenton** [comment]
> you still need to have the checking if you ever want to take it from user input.

Yes, you'd check once and be done rather than every function dealing with paths having to act defensively (and not forget!).

> also, `c:foo` is a trivial path on linux but not on windows.

If a `RelativePath` cannot be resolved relatively then it's up to filesystem functions to handle platform-specific issues (e.g. `c:foo` could be converted to `.\c:foo` on Windows). Alternatively a conversion to `Path` could simply error.

**Comment by traviscross** [comment]
Another possibility that came to mind: `extends_only` (or `only_extends`).

**Comment by dtolnay** [comment]
We discussed this API proposal in today's standard library API team meeting and agreed to accept under the alternative name `is_monotonic`. Monotonic means "only varies in one direction", in contrast with paths like `a/..` and `../b` which contain both ascending and descending components, and paths like `/a/b` which involve ascending all the way to the root before descending. The documentation will need to clarify that the method refers to monotonically *descending* paths, i.e. a path like `../..` would nevertheless return false.

Separately, we are open to considering another ACP that would express variation(s) of this behavior via an API on `Component`, such as `path.components().all(Component::is_normal)`.
